src/editor/editorinteractive.cc

```diff
@@ -951,6 +1287,621 @@ void EditorInteractive::map_changed(const MapWas& action) {
+				return "remove";
+			}
+			NEVER_HERE();
+		};
+
+		std::set<const Widelands::Economy*> economies_to_save;
+		{
+			std::set<Widelands::Serial> saved_mos;
+			for (size_t map_index = map.max_index(); map_index; --map_index) {
+				const Widelands::Coords coords = map.coords(map_index - 1);
+				const Widelands::Field& f = map[map_index - 1];
+				if (f.get_immovable()) {
+					bool skip = false;
+					if (saved_mos.count(f.get_immovable()->serial())) {
+						skip = true;
+					} else {
+						upcast(const Widelands::Building, bld, f.get_immovable());
+						if (bld && bld->get_position() != coords) {
+							skip = true;
+						}
+					}
+					if (!skip) {
+						saved_mos.insert(f.get_immovable()->serial());
+						const Widelands::MapObjectType type = f.get_immovable()->descr().type();
+						switch (type) {
+						case Widelands::MapObjectType::FLAG: {
+							const Widelands::Flag& flag =
+							   dynamic_cast<const Widelands::Flag&>(*f.get_immovable());
+							if (!economies_to_save.count(flag.get_economy(Widelands::wwWARE)))
+								economies_to_save.insert(flag.get_economy(Widelands::wwWARE));
+							if (!economies_to_save.count(flag.get_economy(Widelands::wwWORKER)))
+								economies_to_save.insert(flag.get_economy(Widelands::wwWORKER));
+							write("%s = %s:place_flag(%s:get_field(%u, %u), "
+							      "true)",
+							      var, pvar[flag.owner().player_number() - 1].c_str(),
+							      autogen_map_name.c_str(), coords.x, coords.y);
+							if (flag.current_wares()) {
+								std::map<std::string, unsigned> wares;
+								for (const Widelands::WareInstance* w : flag.get_wares()) {
+									auto it = wares.find(w->descr().name());
+									if (it == wares.end())
+										wares[w->descr().name()] = 1;
+									else
+										++it->second;
+								}
+								write("%s:set_wares({", var);
+								for (const auto& pair : wares)
+									write("   %s = %u,", pair.first.c_str(), pair.second);
+								fw.print_f("   })\n");
+							}
+						} break;
+						case Widelands::MapObjectType::ROAD:
+						case Widelands::MapObjectType::WATERWAY: {
+							const Widelands::RoadBase& road =
+							   dynamic_cast<const Widelands::RoadBase&>(*f.get_immovable());
+							const Widelands::Road* r = dynamic_cast<const Widelands::Road*>(&road);
+							fw.print_f("   %s = %s:place_road(\"%s\", "
+							           "%s:get_field(%u, %u).immovable",
+							           var, pvar[road.owner().player_number() - 1].c_str(),
+							           r ? r->is_busy() ? "busy" : "normal" : "waterway",
+							           autogen_map_name.c_str(),
+							           road.get_flag(Widelands::RoadBase::FlagStart).get_position().x,
+							           road.get_flag(Widelands::RoadBase::FlagStart).get_position().y);
+							const size_t n = road.get_path().get_nsteps();
+							for (size_t i = 0; i < n; ++i) {
+								const char* dir = nullptr;
+								switch (road.get_path()[i]) {
+								case Widelands::WALK_E:
```

Comment:
We have conversion functions in the opposite direction for the waking dirs in `lua_map.cc` and `soldier.cc`. It would be nice to have conversion functions in `walkingdir.h` instead.